test/SILOptimizer/specialize_opaque_type_archetypes.swift

```diff
@@ -0,0 +1,74 @@
+// RUN: %target-swift-frontend -module-name A -enforce-exclusivity=checked -Osize -emit-sil  %s | %FileCheck %s
+public protocol P {
+  func myValue() -> Int64
+}
+
+extension Int64: P {
+  public func myValue() -> Int64 {
+    return self
+  }
+}
+
+@inline(never)
+func useP<T: P> (_ t: T) {
+  print(t)
+}
+
+public func bar(_ x: Int64) -> some P {
+  return x
+}
+
+public func foo(_ x: Int64) -> some P {
+  if x > 0 {
+    return bar(x + 1)
+  }
+
+  return bar(x - 1)
+}
+
+@inline(never)
+func getInt() -> Int64 {
+  return 2
+}
+
+@inline(never)
+func identity<T>(_ t: T) -> T {
+  return t
+}
+
+// CHECK-LABEL: sil @$s1A10testFooBaryyxAA1PRzlF : $@convention(thin) <T where T : P> (@in_guaranteed T) -> () {
```

Comment:
Can you add a few more test cases?
- concrete type is address-only
- the protocol is class-constrained
- the opaque type is in another module (so it can't be specialized)
- assignment to a stored property whose type is opaque (use the typealias/associated type trick I showed you in the earlier comment). I believe this will crash right now because you have to insert a cast, just like you do for function arguments